File: cpp/strings/cadenas.cpp
#include <iostream>
using namespace std;

/*
 * @author   Jorge Eduardo Ascencio Espíndola
 * @version  v1.0
 */

/*
 * @return una cadena con un saludo
 */
string hola(){
  string hello = "Hola";
  return hello;
}

/* 
 * @param word : string
 * 
 * Verifica si la cadena word es palindromo
 * 
 * @return true (1) si  lo es, false (0) en otro caso
*/
bool palindromo(const string& word){
  int f=0, l=word.length()-1; // tamaño de cadena
  while(f<l){
    if(word[f++] != word[l--]) return false; // acceso a cadenas en un indice
  }
  return true;
}

/*
 * @param ch : char
 * @param str : string
 *
 * Cuenta el número de repeticiones del caracter ch en la cadena str
 *
 * @return el número de repeticiones
*/
int repeticiones(const char ch, const string& str){
  int c = 0;
  for(int i=0; i < str.length(); i++){
    if(str[i] == ch) c++; // comparar caracteres
  }
  return c;
}

/*
 * @param word : string
 * 
 * Invierte la cadena word
 *
 * @return la cadena invertida
*/
string invertir(const string& word){
  int l = word.length();
  char inv_str[l]; // crear cadena como arreglo de caracteres
  int f = l-1;
  for(int i=0; i<l;i++){
    inv_str[i] = word[f--]; // 'concatenar' cadenas
  }
  return inv_str;
}
  
int main(){
  string hello = hola();
  cout << hello << "\n";
  
  string word;
  cout << "Escribe una palabra    "; // imprimir en consola
  cin >> word; // obtener entrada desde consola
  cout << "¿Es palindromo?    " << palindromo(word) << "\n";

  char ch;
  cout << "Caracter a buscar   ";
  cin >> ch;
  cout << "repeticiones de " << ch << " : " << repeticiones(ch, word) << "\n";

  string inv_str = invertir(word);
  cout << "cadena invertida   " << inv_str << "\n"; // concatenar cadenas para imprimir
  return 1;
}

File: _experimental/_elixir/cadenas/cadenas.exs
defmodule Cadenas do
  @moduledoc """
  Módulo para saber cómo tratar con cadenas en elixir

    - author:   Jorge Eduardo Ascencio Espíndola
    - version:  v1.0
  """

  @doc """
  Imprime un mensaje de bienvenida
  
  ## Return
     
     la palabra "Hola"
  """
  def hola() do
    "Hola"
  end

  @doc """
  Verifica si una palabra en palindromo

  ## Parameters

    - word: palabra a verificar.

  ## Return 

     true si es palindromo, false en otro caso
  """
  @spec is_palindromo(String.t()) :: Boolean.t()
  def is_palindromo(word) do
    if(String.length(word) > 2) do
      if(String.first(word) != String.last(word)) do
        false
      else
        is_palindromo(String.slice(word, 1..-2))
      end
    else
      String.first(word) == String.last(word)
    end
  end

  @doc """
  Cuenta el número de repeticiones de un cáracter en una cadena

  ## Parameters

    - ch: cáracter a buscar
    - word: palabra a iterar
    - count: número de repeticiones al momento

  ## Return

     El número de repeticiones en toda la palabra
  """
  @spec repeticiones(String.t(), String.t(), Integer.t()) :: Integer.t()
  def repeticiones(ch, word, count) do
    if(String.length(word) > 0) do
      if(String.first(word) == ch) do
        repeticiones(ch, String.slice(word, 1..-1), count + 1)
      else
        repeticiones(ch, String.slice(word,1..-1), count)
      end
    else
      count
    end
  end

  @doc """
  Invierte una cadena

  ## Parameters

    - word: cadena a invertir

  ## Return
     
     La cadena invertida
  """
  @spec reversa(String.t()) :: String.t()
  def reversa(word) do
    String.reverse(word)
  end
end

IO.puts Cadenas.hola()
word = IO.gets "Escribe una palabra   "
word = String.trim(word)
IO.puts "Es palindromo : " <> to_string(Cadenas.is_palindromo(word))
ch = IO.gets "cáracter a buscar   "
ch = String.slice(String.trim(ch), 0..1)
IO.puts "repeticiones : " <> to_string(Cadenas.repeticiones(ch, word, 0))